### PR TITLE
Make P/INS momentum equation implementation more flexible

### PR DESCRIPTION
**Pull request by lingzou**
Both the convection and pressure gradient terms are flexible to use integration by parts or not.

Refs #23121 

## Reason

In current implementation, the momentum equation kernel/BC uses a single flag `conservative_form` to control two things, the first one if the conservative form formula is used for the convection term, and the other if integration by parts should be used for the pressure gradient term; which are really two independent concepts.

## Design

An additional flag `p_int_by_parts` is added in both kernel/BC of the momentum equation, dedicated to control if integration by parts should be applied for the pressure gradient term. The `conservative_form` is kept for the convection term.

## Impact

The code is more flexible to fit different physics simulation needs.

A new test case, `test/tests/finite_element/pins/channel-flow/pressure_gradient.i` is added to test the combination of `conservative_form = false` and `p_int_by_parts = true` condition.

An existing test case, `test/tests/finite_element/pins/channel-flow/pm_friction.i` already demonstrates the `conservative_form = true ` and `p_int_by_parts = true` combination.

### CODE REVIEW

**Comment by lindsayad** [comment]
This would be good for @rui-hu to get a review in

**Comment by rui-hu** [comment]
@lingzou, this PR causes failures of a large number of SAM test cases. Please look into it.

**Comment by lingzou** [comment]
> 

@rui-hu A 'draft' MR has been made in SAM to address the failed teat cases. As soon as this PR is accepted, the SAM MR can be tested and reviewed.

**Comment by rui-hu** [comment]
@lindsayad, this PR looks good to me. The companion draft MR in SAM will go through the review process on this MOOSE PR is accepted. 

I approve this PR.

**Review by rui-hu — DISMISSED** [review]

File: modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C [hunk 1]
```diff
@@ -20,6 +20,9 @@ INSFEFluidMomentumBC::validParams()
 {
   InputParameters params = INSFEFluidIntegratedBCBase::validParams();
   params.addClassDescription("Specifies flow of momentum through a boundary");
+  params.addParam<bool>("conservative_form", false, "if conservative form is used");
```
Comment (GiudGiud) [inline modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C:23]:
```suggestion
  params.addParam<bool>("conservative_form", false, "Whether the conservative form is used");
```

File: modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C [hunk 2]
```diff
@@ -132,32 +140,34 @@ INSFEFluidMomentumBC::computeQpOffDiagJacobian(unsigned int jvar)
   unsigned m = this->mapVarNumber(jvar);
   RealVectorValue vec_vel(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
   Real porosity = _has_porosity ? _porosity[_qp] : 1;
+  Real jac = 0;
 
-  // this is the jacobian w.r.t branch pressure
```
Comment (GiudGiud) [inline modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C:136]: it was nice having it before the switch, you could return earlier

File: modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h [hunk 3]
```diff
@@ -31,6 +34,8 @@ class INSFEFluidMomentumBC : public INSFEFluidIntegratedBCBase
   virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
 
 private:
+  bool _conservative_form;
```
[the remark this answers is not in the document]
Comment (lingzou) [reply modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h]: Updated in the latest commits.

File: modules/navier_stokes/include/kernels/INSFEFluidMomentumKernel.h [hunk 4]
```diff
@@ -29,5 +29,6 @@ class INSFEFluidMomentumKernel : public INSFEFluidKernelStabilization
 
   const VariableGradient & _grad_eps;
   bool _conservative_form;
+  bool _p_int_by_parts;
```
[the remark this answers is not in the document]
Comment (lingzou) [reply modules/navier_stokes/include/kernels/INSFEFluidMomentumKernel.h]: Updated in the latest commits.

File: modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C (hunk 2 above)
Comment (lingzou) [reply modules/navier_stokes/src/bcs/INSFEFluidMomentumBC.C:136]: Updated in the latest commits.

**Review by GiudGiud — APPROVED** [review]

**Review by rui-hu — APPROVED** [review]

**Pull request merged**